File: Neetcode 150/Sliding Window/longest-repeating-character-replacement.cpp

// https://leetcode.com/problems/longest-repeating-character-replacement/description/

class Solution {
public:
    int characterReplacement(string s, int k) {
        
        int maxLength = -0;
        int size = s.size();
        if(size == 0) return maxLength;

        int left = 0, right = 0, maxFreq = 0;
        unordered_map<char, int> mp;
        while(right < size){
            mp[s[right]]++;
            maxFreq = max(maxFreq, mp[s[right]]);

            // (right-left+1) = window size
            // invalid window 
            while((right-left+1) - maxFreq > k){
                mp[s[left]]--;
                left++;
            }

            maxLength = max(maxLength, (right-left+1));
            right++;
        }

        return maxLength;
    }
};

File: Neetcode-150-js/Sliding Window/longest-repeating-character-replacement.js

// https://leetcode.com/problems/longest-repeating-character-replacement/description/

/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {
    const len = s.length;
    if(len === 0) return 0;

    const map = {};
    let left = 0, right = 0, maxFreq = 0, ans = 0;
    while(right < len){
        map[s[right]] = (map[s[right]] === undefined) ? 1 : map[s[right]]+1;
        maxFreq = Math.max(maxFreq, map[s[right]]);

        if((right-left+1)-maxFreq > k){
            map[s[left]]--;
            left++;
        }

        ans = Math.max(ans, (right-left+1));
        right++;
    }

    return ans;
};